### PR TITLE
RFC: hostapd: extend module to allow multiple APs.

### PR DESCRIPTION
**Pull request by qolii**
###### Motivation for this change
It is useful to be able to run multiple APs on a hostapd machine. The traditional ways to do this appear to be:

1. Run multiple instances of the daemon.
2. Run one instance of the daemon with multiple separate config files.

This change implements (1). I prefer this a bit over (2), because it gives each AP its own separate log in journald, and also means you can change configuration and restart just one AP at a time. I have found this to be more flexible in use.

I also have a version of this change implementing (2), if people feel strongly about that.

NOTE: This is a breaking change to the configuration format. My approach has been to take the existing single-AP configuration style, define a submodule around it, and modify the nixos module to take a list of this submodule called `APs`, as follows:
```
{
  enable = true;
  APs = [
    # 5GHz
    { interface = "wlp7s0";
      driver = "nl80211";
      ssid = "acbdef";
      hwMode = "a";
      wpa = true;
      wpaPassphrase = "12345";
      channel = 0;
      extraConfig = ''
        ...
      '';
    }
    # 2.4GHz
    { interface = "wlp5s0";
      driver = "nl80211";
      ssid = "abcdef";
      hwMode = "g";
      wpa = true;
      wpaPassphrase = "12345";
      channel = 0;
      extraConfig = ''
        ...
      '';
    }
  ];
}
```
This will result in two systemd service files, namely, `hostapd-wlp7s0.service` and `hostapd-wlp5s0.service`.

This works well, but will not accept any working existing configurations. I also could not figure out how to extend the hostapd entry in `rename.nix`. I expect these things are a big problem for the acceptance of this change, so if anyone has advice on how to achieve this more gently, please let me know.

###### systemd ordering
Upon making this change, I noticed failures at startup due to systemd ordering, because hostapd would try to start too early. I have thus reworked the ordering a little bit, and observe a completely reliable service on my machine. However, I am not an expert in all the different types of ways to do this, so any advice on this front is also very welcome.

In summary, this change has been very functional for me in testing.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by qolii** [comment]
[Mentioning @wkennington and @Phreedom, as folks in the maintainers list for hostapd.]

**Comment by qolii** [comment]
A thought: given that we are faced with a breaking change anyway, would it be worthwhile to add a `bridge` field to the submodule, and then have hostapd service unit `bindsTo` and `after` the bridge's device unit? I don't know if there is currently an actual dependency on that, but I think there should be.

**Comment by qolii** [comment]
Well, I went for it. This actually lets me simplify the dependency list now that the bridge dependency is explicit. I rebooted my machine a bunch of times an observed stable bring-up of everything.

Since it's a separate commit, it will be easy to back out if there are issues.

**Comment by flokli** [comment]
@qolii Thanks for the PR! Sorry this went unnoticed for so long.

I gave this a try together with @NinjaTrappeur. We rebased your commits on the latest master, a more recent version of the branch can be found at https://github.com/flokli/nixpkgs/commits/hostapd-multiple-ifs.

I incorporated the systemd interface escaping that was done in master since your PR.

I do have more comments about it. Do you still want to merge this in? If yes, could you reset your PR to that version, so I can comment in here?

**Comment by qolii** [comment]
@flokli, thanks for getting back to me! This sounds great. Please give me another day or two, I'm a bit swamped at the moment.

**Comment by flokli** [comment]
sure :-)

**Comment by qolii** [comment]
Hi @flokli, so, what I did was remade by the branch on my fork by cherry-picking the two commits from the branch on your fork. So I think those are now what you see here.

Is this what you had in mind?

If so, comment away! What are you thinking?

**Comment by qolii** [comment]
(And just fixed merge conflicts)

**Comment by qolii** [comment]
(and just reinstated the `noScan` option that got stomped)

**Comment by qolii** [comment]
@flokli / @NinjaTrappeur, any thoughts?

**Comment by flokli** [comment]
@qolii this doesn't look too bad to me, but for the next two weeks, I won't have access to the hardware available to test it.

I remember @NinjaTrappeur mentioning a lot of hostapd options were still missing. Instead of adding all these, do you think it might make sense to refactor this to something like https://github.com/Infinisil/rfcs/blob/config-option/rfcs/0042-config-option.md describes?

**Comment by qolii** [comment]
@floli, I like the idea of doing a refactor. However, would you be amenable to merging this first (as I think this provides genuinely useful functionality), and doing that in a separate PR?

I'm happy to give the refactor a shot myself.

File: nixos/modules/services/networking/hostapd.nix
Comment (flokli) [inline nixos/modules/services/networking/hostapd.nix]: It's probably not too complicated to add code for this inside this PR.

File: nixos/modules/services/networking/hostapd.nix
Comment (flokli) [inline nixos/modules/services/networking/hostapd.nix]: This could be moved into the generic `services.hostapd` description.

File: nixos/modules/services/networking/hostapd.nix
Comment (flokli) [inline nixos/modules/services/networking/hostapd.nix]: probably the same applies for this one as well - In `networking.supplicant.<name>`, `<name>` can be separated by spaces, so matching automatically would be a bit more complicated.

File: nixos/modules/services/networking/hostapd.nix
Comment (flokli) [inline nixos/modules/services/networking/hostapd.nix]: This should be singular.

File: nixos/modules/services/networking/hostapd.nix
Comment (flokli) [inline nixos/modules/services/networking/hostapd.nix]: I'm not sure if I'm so happy with the `APs = []` part - it's quite unintuitive compared to how the rest of the module system looks like, and `hostapd` too only knows about bssids and interfaces, not `APs`.

Can we make this similar to `networking.supplicant`? Having the description in `services.hostapd`, and a `services.hostapd.<(interface_)name>.*`, like `networking.supplicant.<name>.*`?

From my understanding, there can only be one `hostapd` process per physical interface, and configuring multiple bssids would be part of the per-interface config file, which would need to duplicated for dual/multi-band setups.

**Review by flokli — CHANGES_REQUESTED** [review]
After reading through the `hostapd.conf` manpage once more, some comments.

The changed options also require a changelog entry.

**Comment by ivan** [comment]
For the author, reviewers, and committers: this PR was scanned and appears to add a use of the deprecated `types.string`, which emits a warning as of #66346. Before merging, please change this to another type, possibly:

- `types.str` for a single string where merging does not make sense, or cannot work
- `types.lines` for multi-line configuration or scripts where merging is possible
- `types.listOf types.str` for a mergeable list of strings

**Comment by flokli** [comment]
@qolii the usage of `types.string` was fixed in master in 478e7184f88db1364cc75107036f7c4decc0cc41.

I'd be more in favor of doing the refactoring inside this PR - we introduce a bunch of new options / changes, which we'd have to worry about deprecating later again.

**Comment by qolii** [comment]
Still important to me! I'll try to rework this based on the comments so far. I know has changed in Nix-land since I first proposed this anyway.

**Comment by RaitoBezarius** [comment]
Can I help you to get this PR merged?

**Comment by AndersonTorres** [comment]
Last activity from @qolii is from 2020:

https://github.com/NixOS/nix/pull/3166

Closing as dead.

**Pull request closed without merging**
